src/runtime/linux/auxvec.rs

```diff
@@ -0,0 +1,86 @@
+//! Reads /proc/self/auxv on Linux systems
+
+use std::prelude::v1::*;
+use std::slice;
+use std::mem;
+use std::fmt;
+
+/// Simple abstraction for the ELF Auxiliary Vector
+///
+/// the elf.h provide the layout of the single entry as auxv_t.
+/// The desugared version is a usize tag followed by a union with
+/// the same storage size.
+pub struct AuxVec {
+    raw: Vec<usize>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[allow(dead_code)]
+/// ELF Auxiliary vector entry types
+///
+/// The entry types are specified in `linux/auxvec.h`.
+pub enum AT {
+    /// CPU Hardware capabilities, it is a bitfield.
+    HWCAP = 16,
+    /// CPU Hardware capabilities, additional bitfield.
+    HWCAP2 = 26,
+}
+
+impl fmt::Debug for AuxVec {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_list()
+            .entries(self.raw.chunks(2).map(|kv| (kv[0], kv[1])))
+            .finish()
+    }
+}
+
+impl AuxVec {
+    /// Reads /proc/self/auxv into AuxVec.
+    // TODO: make it less ugly if possible
+    pub fn new() -> Result<Self, ::std::io::Error> {
+        use std::io::Read;
+        let mut file = ::std::fs::File::open("/proc/self/auxv")?;
```

Comment:
This is a good first step, but we should only do this if libc's `getauxval` does not exist. I think we can add `getauxval` as a `weak_symbol`, and check whether the value of a function pointer to it is null or not at run-time. If it is not null we should just call that, and if it is null, then the binary does not have a linked extern C `getauxval`, and we can proceed to read `proc/self/auxv`.

I wish someone (@alexcrichton , @BurntSushi ) would review #211 since that would allow us to add some more ARM targets for testing and use qemu-system instead of qemu-user to test this on linux and android targets.

Reply:
I avoided `getauxval` since it is nonstandard, I wouldn't rely on weak_symbol to check for its presence (static linking and non-gnu libcs aren't uncommon:))

Reply:
Typically we don't use weak symbols as they're not very portable, but a solution is to do what libstd does which is `dlsym` at runtime.

Reply:
IIRC the problem was that `/proc/self/auxv` is not guaranteed to exist on all Android targets, but `libc::getauxval` is (and also, `libc::getauxval` is always faster than reading `/proc/self/auxv` on all platforms that have it). 

Anyways, we don't need to add `libc::getauxval` on this PR, that was more me thinking out loud. This PR is a good step forward as it is. Good job!